PEP 394: Allow for more flexibility in handling /usr/bin/python (#989)
python/peps: pep-0394.txt

Author: Kerrick Staley <mail@kerrickstaley.com>,
        Nick Coghlan <ncoghlan@gmail.com>,
        Barry Warsaw <barry@python.org>,
        Petr Viktorin <encukou@gmail.com>,
        Miro Hrončok <miro@hroncok.cz>,
Status: Active
Type: Informational
Content-Type: text/x-rst
Created: 02-Mar-2011
Post-History: 04-Mar-2011, 20-Jul-2011, 16-Feb-2012, 30-Sep-2014, 28-Apr-2018,
              12-Apr-2019
Resolution: https://mail.python.org/pipermail/python-dev/2012-February/116594.html

Abstract
========

This PEP provides a convention to ensure that Python scripts can continue to
be portable across ``*nix`` systems, regardless of the version of the Python
interpreter invoked by the ``python`` command.

* ``python2`` will refer to some version of Python 2.x.
* ``python3`` will refer to some version of Python 3.x.
* Distributions can choose whether ``python`` is installed and, if it is,
  whether it refers to the same target as ``python2`` or ``python3``.
* End users should be aware that the ``python`` command is inconsistent
  across Unix-like systems.
* In activated virtual environments, ``python`` always refers to the
  environment's Python version.

> * Upstream projects should prefer the ``python3`` shebang even if they support

Author: 4a4cdf7c

  both Python 3 and Python 2.
* Upstream projects only supporting Python 2 should use the ``python2``
  shebang.

Recommendation
==============

For distributors
----------------

* Unix-like software distributions (including systems like macOS and
  Cygwin) should install the ``python2`` command into the default path
  whenever a version of the Python 2 interpreter is installed, and the same
  for ``python3`` and the Python 3 interpreter.
* When  invoked, ``python2`` should run some version of the Python 2
  interpreter, and ``python3`` should run some version of the Python 3
  interpreter.
* If the ``python`` command is installed, it should either invoke the same
  version of Python as the ``python3`` command or as the ``python2``
  command.
* Distributors may choose to set the behavior of the ``python`` command,
  not provide it at all, or make it configurable by the user or
  system administrator.
* The Python 3.x ``idle``, ``pydoc``, and ``python-config`` commands should
  likewise be available as ``idle3``, ``pydoc3``, and ``python3-config``

>   (and similarly for Python 2.x). The original commands should either invoke

Reviewer: "and similarly for Python 2.x" is a bit ambiguous.  Are you suggesting that if distributors include these tools for Python 2, that they be named `idle2`, `pydoc2`, and `python2-config`?  Perhaps it's better to be explicit here.

Author: abf3bf3

  when Python 3.x is not yet supported, or changing ``python3`` shebangs to
  ``python3.8`` if the software is built with Python 3.8.
* When a virtual environment (created by the PEP 405 ``venv`` package or a
  similar tool such as ``virtualenv`` or ``conda``) is active, the ``python``
  command should refer to the virtual environment's interpreter and should
  always be available.
  The ``python3`` or ``python2`` command (according to the environment's
  interpreter) should also be available.

For developers
--------------

* In order to tolerate differences across platforms, new code that needs
  to invoke the Python interpreter should not specify ``python``, but rather
  should specify either ``python3`` or ``python2`` (or the more specific
  ``python3.x`` and ``python2.x`` versions; see the `Migration Notes`_).
  This distinction should be made in shebangs, when invoking from a shell
  script, when invoking via the system() call, or when invoking in any other
  context.
* When reinvoking the interpreter from a Python script, querying
  ``sys.executable`` to avoid hardcoded assumptions regarding the
  interpreter location remains the preferred approach.

> * Scripts that are deliberately written to be source compatible with both

Reviewer 2: I think we need to present additional options here, as our actual advice to developers is "don't rely on hardcoded shebang lines, and instead let tools rewrite them for you".

That means the preferred option is "Use a Python package installer that supports shebang line rewriting, like `pip install --user` or `pipx install`, rather than copying around plain text files" (perhaps with a reference to https://packaging.python.org/guides/installing-stand-alone-command-line-tools/ ).

Option 2 is "Use the environment-aware `/usr/bin/env python` shebang line, and require that the script be executed inside an activated Python virtual environment"

Option 3 is "Use the generic `/usr/bin/env python` shebang line, but ensure that will resolve to a suitable interpreter on the systems that you're targeting if the script needs to run outside a virtual environment". Adding in `ln -s ~/.local/bin/python /usr/bin/python3` or `sudo ln -s /usr/local/bin/python /usr/bin/python3` for the target systems that you care about is a much lower impact change than rewriting all your shebang lines to add a version qualifier (especially since making that change will break the scripts on systems that only have Python 2).

The only time that using the `python3` shebang genuinely makes sense is when the script wouldn't run on Python 2 anyway - in those cases, using the more specific shebang only increases your target platform compatibility, it never reduces it.

Reviewer 2: If we don't want that level of complexity in the PEP, than I think we should make the guidance explicitly:

- `/usr/bin/env python` for source compatible scripts, or Python 3 scripts that are unlikely to ever be exposed to target environments where that invocation won't work (i.e. Python 3 only systems when outside a virtual environment)
- `/usr/bin/env python3` for scripts that require Python 3 or later
- `/usr/bin/env python2` for scripts that don't support Python 3 at all
- where possible, use a Python package or script installer (e.g. `pip` or `pipx`) instead of copying scripts directly to a target system, as they are able to rewrite shebang lines to suit the target environment

Hardcoding specific paths to Python binaries should be restricted to non-portable scripts that only target particular platforms and don't need to respect virtual environments.

Reviewer 2: After discussing this with @encukou in person, I created a PR against the PR branch to better convey what I'm proposing here: https://github.com/hroncok/peps/pull/1/files

``python2`` command by default, there was previously no way for Python 2 code
(or any code that invokes the Python 2 interpreter directly rather than via
``sys.executable``) to reliably run on all Unix-like systems without
modification, as the ``python`` command would invoke the wrong interpreter
version on some systems, and the ``python2`` command would fail completely
on others. This PEP originally provided a very simple mechanism
to restore cross-platform support, with minimal additional work required
on the part of distribution maintainers. Simplified, the recommendation was:

1. The ``python`` command was preferred for code compatible with both
   Python 2 and 3 (since it was available on all systems, even those that
   already aliased it to Python 3).
2. The ``python`` command should always invoke Python 2 (to prevent
   hard-to-diagnose errors when Python 2 code is run on Python 3).
3. The ``python2`` and ``python3`` commands should be available to specify
   the version explicitly.

However, these recommendations implicitly assumed that Python 2 would always be
available. As Python 2 is nearing its end of life in 2020 (PEP 373, PEP 404),

> distributions are making Python 2 optional or removing entirely.

Author: 4a4cdf7c

This means either removing the ``python`` command or switching it to invoke
Python 3, invalidating respectively the first or second recommendation.
Also, some distributors decided that their users are better served by

> ignoring the PEP's recommendations, making the PEP's supposedly

Author: 4a4cdf7c

Author: Kerrick Staley <mail@kerrickstaley.com>,
        Nick Coghlan <ncoghlan@gmail.com>,
        Barry Warsaw <barry@python.org>,
        Petr Viktorin <encukou@gmail.com>,
        Miro Hrončok <miro@hroncok.cz>,
Status: Active
Type: Informational
Content-Type: text/x-rst
Created: 02-Mar-2011
Post-History: 04-Mar-2011, 20-Jul-2011, 16-Feb-2012, 30-Sep-2014, 28-Apr-2018,
              12-Apr-2019
Resolution: https://mail.python.org/pipermail/python-dev/2012-February/116594.html

Abstract
========

This PEP provides a convention to ensure that Python scripts can continue to

> be portable across ``*nix`` systems, regardless of the version of the Python

Reviewer 3: As a point of clarification, are we considering MacOS as a ``*nix`` system?

Reviewer: Pedantically, it's "macOS" but I hope the answer is "yes".

Author: Let me know if you want e to explicitly put this information in.

> be portable across ``*nix`` systems, regardless of the version of the Python

Reviewer 3: Perhaps:

This PEP outlines the behavior of Python scripts when the ``python`` command is invoked. Depending on a distribution or system configuration, ``python`` may or may not be installed. If ``python`` is installed its target interpreter may refer to ``python2`` or ``python3``. End users may be unaware of this inconsistency across Unix-like systems. This PEP's goal is to reduce user confusion about what ``python`` references and what will be the script's behavior.

The recommendations in the next section of this PEP will outline the behavior when:
- using virtual environments
- cross-platform scripts with shebangs for either ``python2`` or ``python3`` 

The PEP's goal is to clarify the behavior for script end users, distribution providers, and script maintainers / authors.

interpreter invoked by the ``python`` command.

* ``python2`` will refer to some version of Python 2.x.
* ``python3`` will refer to some version of Python 3.x.
* Distributions can choose whether ``python`` is installed and, if it is,
  whether it refers to the same target as ``python2`` or ``python3``.
* End users should be aware that the ``python`` command is inconsistent
  across Unix-like systems.
* In activated virtual environments, ``python`` always refers to the
  environment's Python version.
* Cross-platform scripts should prefer the ``python3`` shebang even if they
  support both Python 3 and Python 2.
* Cross-platform scripts only supporting Python 2 should use the ``python2``
  shebang.

Recommendation
==============

> For distributors

Reviewer 3: ```suggestion

Our recommendations are detailed below. We call out any expectations that these recommendations are based upon.

For distributors
```

----------------

> * Unix-like software distributions (including systems like macOS and

Reviewer 3: ```suggestion
* We expect Unix-like software distributions (including systems like macOS and
```

  Cygwin) should install the ``python2`` command into the default path
  whenever a version of the Python 2 interpreter is installed, and the same
  for ``python3`` and the Python 3 interpreter.
* When  invoked, ``python2`` should run some version of the Python 2
  interpreter, and ``python3`` should run some version of the Python 3
  interpreter.

> * If the ``python`` command is installed, it should either invoke the same

Reviewer 3: ```suggestion
* If the ``python`` command is installed, it is expected to invoke either the same
```

  version of Python as the ``python3`` command or as the ``python2``
  command.

> * Distributors may choose to set the behavior of the ``python`` command,

Reviewer 3: ```suggestion
* Distributors may choose to set the behavior of the ``python`` command as follows:
    - ``python2``,
    - ``python3``,
    - not provide ``python`` command
    - allow ``python`` to be configurable by an end user or a system administrator.
```

Author: (Note for self: several lines bellow need to be removed if this goes in).

  not provide it at all, or make it configurable by the user or
  system administrator.
* The Python 3.x ``idle``, ``pydoc``, and ``python-config`` commands should
  likewise be available as ``idle3``, ``pydoc3``, and ``python3-config``;
  Python 2.x versions as ``idle2``, ``pydoc2``, and ``python2-config``.

>   The original commands should either invoke the same version of Python

Reviewer 3: The unspecified commands with no version number should ...

Author: What about simple: The commands with no version number should...

>   The original commands should either invoke the same version of Python

Author: ```suggestion
  The commands with no version number should either invoke the same version of Python
```

  as the ``python`` command, or not be available at all.

> * When packaging software, distributors are encouraged to change less specific

Reviewer 3: ```suggestion
* When packaging third party Python scripts, distributors are encouraged to change less specific
```

>   shebangs to more specific ones. This ensures software is used with the latest

Reviewer 3: ```suggestion
  shebangs to more explicit and specific ones. This encourages scripts to use the latest
```

Author: *encourages* actually makes me thing it somehow makes it more likely. In fact, it just makes it so. I still consider *ensures* or *makes sure* to fit better here.

  version of Python available, and it can remove a dependency on Python 2.

>   The specifics are left on the distributors.

Reviewer 3: ```suggestion
  The details on what specifics to set are left to the distributors; though, example specifics could include: 
```

Author: Note to self: Following lines need to be adjusted if this goes in.

  Examples of this include changing ``python`` shebangs to ``python3``
  when Python 3.x is supported, changing ``python`` shebangs to ``python2``
  when Python 3.x is not yet supported, or changing ``python3`` shebangs to
  ``python3.8`` if the software is built with Python 3.8.
* When a virtual environment (created by the PEP 405 ``venv`` package or a
  similar tool such as ``virtualenv`` or ``conda``) is active, the ``python``
  command should refer to the virtual environment's interpreter and should
  always be available.
  The ``python3`` or ``python2`` command (according to the environment's

>   interpreter) should also be available.

Reviewer 3: ```suggestion
  interpreter version) should also be available.
```

> For developers

Reviewer 3: Do we want to split this into script developers/maintainers and script users (who may not be the script's author)?

Reviewer 2: There aren't any recommendations for script users (unless you count "You may need to edit shebang lines to get scripts to work on your system" as a recommendation)

Reviewer 3: I think that it's worth, even though the recommendation is short, to split out script end users from script developers.

--------------

* When reinvoking the interpreter from a Python script, querying
  ``sys.executable`` to avoid hardcoded assumptions regarding the
  interpreter location remains the preferred approach.
* While far from being universally available, ``python`` remains the
  preferred spelling for explicitly invoking Python, as this is the
  spelling that virtual environments make consistently available
  across different platforms and Python installations.

> * In shebang lines, the preferred spelling is ``/usr/bin/env python``,

Author: Note that we proposed this PR without this recommendation so I leave the decision to you and @ncoghlan. I'd happily keep it or revert it back as you see fit.

> * In shebang lines, the preferred spelling is ``/usr/bin/env python``,

Reviewer 3: This case should be combined with the case on Line 104.

> * In shebang lines, the preferred spelling is ``/usr/bin/env python``,

Reviewer: I want to say that a general `/usr/bin/env python` recommendation outside of venvs is very dangerous.  When I was working on Debian/Ubuntu, such things lead to mysterious problems with serious ramifications.  So we went about actively discouraging such shebangs.

I think we need help from tooling here.  Ideally, end users wouldn't have to worry about it at all, and either the platform builds would modify the shebang (as I *think* but don't really remember any more whether Debuntu does this), or installing into a venv would rewrite the shebang.

So I'm quite concerned that a general recommendation will cause users serious problems.  If we want to make progress on this update and can't agree on shebang recommendations, we should remove it from this update and discuss further in a separate PR, bug, or forum.

> * In shebang lines, the preferred spelling is ``/usr/bin/env python``,

Reviewer 4: @warsaw Is there a difference here between system scripts and those distributed by third parties? If Debian ships a tool that happens to be written in Python (eg /usr/bin/calibre), it makes sense for its shebang to avoid /usr/bin/env, but if I'm going to give people a script that depends on Python, it makes more sense to let their environment select it.

> * In shebang lines, the preferred spelling is ``/usr/bin/env python``,

Reviewer: @Rosuav That's exactly the situation that broke people on Ubuntu.  They'd end up using some random Python they forgot they installed, with an incompatible library pip installed and it broke their script.  Then they blamed the wrong people, when really they should have pointed their script to the right Python executable, or fixed their `$PATH` correctly.  Okay, sure this is user error, but I know with a PEP 394 recommendation to use `/usr/bin/env`, it'll happen more often.  It might even be that they *think* they're running in a venv, but really they aren't.

> * In shebang lines, the preferred spelling is ``/usr/bin/env python``,

Author: Note that the original PEP nor our PR have ever gone into `/usr/bin/python...` vs `/usr/bin/env python...` shebangs. What if we keep it that way and than we can argue about that in another PR?

> * In shebang lines, the preferred spelling is ``/usr/bin/env python``,

Reviewer: +1 - This is the only controversial bit still left IIRC.  I'm happy to defer this and push out an update with the bits we do agree with.  @encukou @ncoghlan Would that be okay with you?

> * In shebang lines, the preferred spelling is ``/usr/bin/env python``,

Reviewer 5: Yes!

> * In shebang lines, the preferred spelling is ``/usr/bin/env python``,

Reviewer 2: Resolved by removing the "preferred" wording, and instead qualifying this bullet point as being specific to scripts that are only expected to be run in virtual environments: https://github.com/python/peps/commit/4b20f8c130df2c6b9b1a9900ea41eca51999ee05

I then removed the two paragraphs about Py2 vs Py3, since they weren't needed any more, and the text flows better without them.

  2 at all, it is recommended to instead use ``python3`` and
  ``/usr/bin/env python3``, as these will never invoke Python 2, and are
  expected to work for both Python 3 virtual environments and Python 3
  system installations
* For Python 2 only scripts that do not support being executed on Python
  3 at all, it is recommended to instead use ``python2`` and
  ``/usr/bin/env python2``, as these will never invoke Python 3, and are
  expected to work for Python 2 virtual environments and at least some
  Python 2 system installations.
* In cases where the script is expected to be executed outside virtual
  environments, developers will need to be aware of the following
  discrepancies across platforms and installation methods:

  * Older Linux distributions will provide a ``python`` command that
    refers to Python 2. Most of these distros do *not* provide a
    ``python2`` command.
  * Some newer Linux distributions will provide a ``python`` command that
    refers to Python 3
  * Some Linux distributions will not provide a ``python`` command at
    all by default, but will provide a ``python3`` command by default

> * When potentially targeting these environments, developers may either

Reviewer 3: These remaining 3 bullets should be moved under the inconclusive ``python`` case as possible resolutions.

Author: Sorry. I cannot read any more, could you please point em to where would you except these 3 bullets points?

on the part of distribution maintainers. Simplified, the recommendation was:

1. The ``python`` command was preferred for code compatible with both
   Python 2 and 3 (since it was available on all systems, even those that
   already aliased it to Python 3).
2. The ``python`` command should always invoke Python 2 (to prevent
   hard-to-diagnose errors when Python 2 code is run on Python 3).
3. The ``python2`` and ``python3`` commands should be available to specify
   the version explicitly.

However, these recommendations implicitly assumed that Python 2 would always be
available. As Python 2 is nearing its end of life in 2020 (PEP 373, PEP 404),
distributions are making Python 2 optional or removing it entirely.
This means either removing the ``python`` command or switching it to invoke
Python 3. Some distributors also decided that their users were better served by
ignoring the PEP's recommendations, and provided system administrators with the
freedom to configure their systems based on the needs of their particular
environment.

.. _rationale:

Current Rationale
=================

As of 2019, activating a Python virtual environment (or its functional

> equivalent) is the best way to obtain a consistent cross-platform and

Reviewer 3: ```suggestion
equivalent) prior to script execution is one way to obtain a consistent cross-platform and
```

> cross-distribution experience.

Reviewer 2: If you're not writing Linux distro packages, you should absolutely be using a venv for everything.

You do need to use `pipx`, and you do need to reinstall them if you change Python versions.

> cross-distribution experience.

Reviewer 2: If you *don't* assume the use of a venv, then everything is awful:

* `python` probably doesn't work on Windows (or runs something random)
* `py` *only* works on Windows
* `python` is only reliably available inside a venv
* `python2` isn't reliably available anywhere
* `python3` doesn't work on Windows

> cross-distribution experience.

Reviewer 2: Or, you run them using an explicit interpreter invocation, and forget about relying on shebang lines.

> cross-distribution experience.

Reviewer: This all feels like it's outside the scope of this PEP.  With this view, we have to endorse a third party tool (`pipx`) and a deployment workflow, and teach people how to use virtual environments (and do they use `virtualenv` or `$python -m venv`?).  The reinstallation issue is problematic too.  I don't know what any of this has to do with our recommendations to *nix distributions.  If we really want to make these recommendations, we should do that in a separate PEP, IMHO.

> cross-distribution experience.

Reviewer: One other thought: If you're using zip applications (pex, shiv), you should absolutely *not* be using venvs or `/usr/bin/env`.

> cross-distribution experience.

Reviewer 2: It's a response to the original draft of this PR, which advised *all* script publishers to unconditionally use `/usr/bin/python3` in their scripts, which I consider to be a bad default recommendation, since it breaks virtual environments.

`/usr/bin/env python` is better generic advice, since it respects virtual environments, and leaves the choice of specifying something else in their shebang line to folks that specifically want to support running outside a venv, and should have a better idea of which kinds of environments they want to target, and hence which compatibility trade-offs they can make in choosing an exact spelling.

Tools like `pipx`, `zipapp`, `shiv` etc all run an installer at some point, so they allow publishers to rely on console script definitions which write the shebang line at install time, rather than needing to rely on a hardcoded shebang line in the source file(s).

The relevance of all this to PEP 394 is that platform providers need to account for publisher behaviour in designing their Python user experience, and that's easier for them to do if our advice to publishers is "assume platforms will either rely on a Python installer, so you can use generated console scripts, or else that they will make themselves look somewhat like a virtual environment".

> Accordingly, it is entirely reasonable for publishers to expect the

Reviewer 3: ```suggestion
Accordingly, publishers can expect users of the software to provide a suitable execution environment.

availability of such an environment to get their software working
correctly, and push responsibility for handling other environments
(such as system Python installations) onto consumers of the software.

> As part of that however, it is also appropriate to give Python

Reviewer 3: I would remove this paragraph.

Author: Kerrick Staley <mail@kerrickstaley.com>,
        Nick Coghlan <ncoghlan@gmail.com>,
        Barry Warsaw <barry@python.org>,
        Petr Viktorin <encukou@gmail.com>,
        Miro Hrončok <miro@hroncok.cz>,
        Carol Willing <willingc@gmail.com>,
Status: Active
Type: Informational
Content-Type: text/x-rst
Created: 02-Mar-2011
Post-History: 04-Mar-2011, 20-Jul-2011, 16-Feb-2012, 30-Sep-2014, 28-Apr-2018,
              12-Apr-2019
Resolution: https://mail.python.org/pipermail/python-dev/2012-February/116594.html

Abstract
========

This PEP outlines the behavior of Python scripts when the ``python`` command
is invoked.
Depending on a distribution or system configuration,
``python`` may or may not be installed.
If ``python`` is installed its target interpreter may refer to ``python2``
or ``python3``.
End users may be unaware of this inconsistency across Unix-like systems.
This PEP's goal is to reduce user confusion about what ``python`` references
and what will be the script's behavior.

The recommendations in the next section of this PEP will outline the behavior
when:

* using virtual environments

> * cross-platform scripts with shebangs for either ``python2`` or ``python3``

Reviewer 5: ```suggestion
* writing cross-platform scripts with shebangs for either ``python2`` or ``python3``
```

  interpreter, and ``python3`` should run some version of the Python 3
  interpreter.
* If the ``python`` command is installed, it is expected to invoke either
  the same version of Python as the ``python3`` command or as the ``python2``
  command.
* Distributors may choose to set the behavior of the ``python`` command
  as follows:

   * ``python2``,
   * ``python3``,
   * not provide ``python`` command,
   * allow ``python`` to be configurable by an end user or
     a system administrator.

* The Python 3.x ``idle``, ``pydoc``, and ``python-config`` commands should
  likewise be available as ``idle3``, ``pydoc3``, and ``python3-config``;
  Python 2.x versions as ``idle2``, ``pydoc2``, and ``python2-config``.
  The commands with no version number should either invoke the same version
  of Python as the ``python`` command, or not be available at all.
* When packaging third party Python scripts, distributors are encouraged to
  change less specific shebangs to more specific ones.
  This ensures software is used with the latest version of Python available,
  and it can remove a dependency on Python 2.
  The details on what specifics to set are left to the distributors;

>   though, example specifics could include:

Reviewer 5: ```suggestion
  though. Example specifics could include:
```

  though. Example specifics could include:

  * Changing ``python`` shebangs to ``python3`` when Python 3.x is supported.
  * Changing ``python`` shebangs to ``python2`` when Python 3.x is not yet
    supported.
  * Changing ``python3`` shebangs to ``python3.8`` if the software is built
    with Python 3.8.

* When a virtual environment (created by the PEP 405 ``venv`` package or a
  similar tool such as ``virtualenv`` or ``conda``) is active, the ``python``
  command should refer to the virtual environment's interpreter and should
  always be available.
  The ``python3`` or ``python2`` command (according to the environment's
  interpreter version) should also be available.

For Python script publishers
----------------------------

* When reinvoking the interpreter from a Python script, querying
  ``sys.executable`` to avoid hardcoded assumptions regarding the
  interpreter location remains the preferred approach.
* Encourage your end users to use a virtual environment.
  This makes the user's environment more predictable (possibly resulting
  in fewer issues), and helps avoid disrupting their system.
* For scripts that are only expected to be run in an activated virtual

>   environment, shebang lines can be written as ``/usr/bin/env python``,

Author: ```suggestion
  environment, shebang lines can be written as ``#!/usr/bin/env python``,
```

  as this instructs the script to respect the active virtual environment.
* In cases where the script is expected to be executed outside virtual
  environments, developers will need to be aware of the following
  discrepancies across platforms and installation methods:

  * Older Linux distributions will provide a ``python`` command that

>     refers to Python 2. Most of these distros do *not* provide a

Reviewer: The second sentence feels a little too imprecise.  What exactly is an "older Linux distribution" and how many of them count as "most"?  Perhaps:

> Older Linux distributions will provide a ``python`` command that refers to Python 2, and will likely not provide a ``python2`` command

    ``python2`` command.
  * Some newer Linux distributions will provide a ``python`` command that

>     refers to Python 3

Author: ```suggestion
    refers to Python 3.
```

  * Some Linux distributions will not provide a ``python`` command at

>     all by default, but will provide a ``python3`` command by default

Author: ```suggestion
    all by default, but will provide a ``python3`` command by default.
```

  Avoiding shebangs (via the console_scripts Entry Points ([9]_) or similar
  means) is the recommended workaround for this problem.
* Applications designed exclusively for a specific environment (such as
  a container or virtual environment) may continue to use the ``python``
  command name.

For end users of Python
-----------------------

* While far from being universally available, ``python`` remains the
  preferred spelling for explicitly invoking Python, as this is the
  spelling that virtual environments make consistently available
  across different platforms and Python installations.
* For software that is not distributed with (or developed for) your system,
  we recommend using a virtual environment, possibly with an environment
  manager like ``conda`` or ``pipenv``, to help avoid disrupting your system
  Python installation.

These recommendations are the outcome of the relevant python-dev discussions
in March and July 2011 ([1]_, [2]_), February 2012 ([4]_),
September 2014 ([6]_), discussion on GitHub in April 2018 ([7]_),
on python-dev in February 2019 ([8]_), and during the PEP update review

> in May 2019 ([10]_).

Author: ```suggestion
in May/June 2019 ([10]_).
```

  a container or virtual environment) may continue to use the ``python``
  command name.

For end users of Python
-----------------------

* While far from being universally available, ``python`` remains the
  preferred spelling for explicitly invoking Python, as this is the
  spelling that virtual environments make consistently available
  across different platforms and Python installations.
* For software that is not distributed with (or developed for) your system,
  we recommend using a virtual environment, possibly with an environment

>   manager like ``conda`` or ``pipenv``, to help avoid disrupting your system

Reviewer 3: Should we mention the built-in `venv`?

Reviewer 2: Perhaps in parentheses in the first bullet point? "... that virtual environments (as created by the standard library's `venv` module or the PyPA's `virtualenv` project) consistently ..."

Reviewer 2: I'm going to leave this for now, as these do get mentioned further up.

  versa. That way, if a sysadmin does decide to replace the installed
  ``python`` file, they can do so without inadvertently deleting the
  previously installed binary.

> * When the Python 2 interpreter becomes uncommon, scripts should nevertheless

Reviewer 2: We should reword this bullet point, since we now expect Python 3 -> Python 4 to be a "enough changes have accumulated since Python 3.0 that they're effectively different languages" kind of event, rather than a Python 3.0 style compatibility break. Something like:

> Even as the Python 2 interpreter becomes less common, scripts should nevertheless continue to use the `python3` convention rather than just `python`. In the event of a future Python 4.0 release, distributors may then choose to simply alias `python3` to the same binary as `python4` (since the degree of change is now expected to be the same as in any other Python feature release), or else they may choose to maintain parallel `python3` and `python4` stacks for a time (depending on the needs of their specific audience).

Author: I'd happy to change this, however is there some authoritative source I can link to? I don't want this PEP to be the place that promises/defines/guesses the plans for Python 4. If not, we could also reword this to work either way. Something like: "When Python 2 is forgotten, still use "python3". If Python 4 will be a disaster, it will help you, OTOH if Python 4 will be peace of cake, you can either alias "python3" to "python4" or decide to keep them parallel installable for a time."

Reviewer 6: There is no authoritative statement beyond some email threads where that seems to be the prevailing feeling towards Python 4.

Author: That's my understanding as well. I don't want this PEP to be used as a reference about Python 4. I'll try to go with a statement that is more universal.

Author: 36e61fab

  continue to use the ``python3`` convention rather that just ``python``. This
  will ease transition in the event that yet another major version of Python
  is released.
* If these conventions are adhered to, it will become the case that the
  ``python`` command is only executed in an interactive manner as a user

>   convenience.

Reviewer 2: extra clause: "... convenience, or else when using a virtual environment or similar mechanism."

Author: 4a4cdf7c

  versa. That way, if a sysadmin does decide to replace the installed
  ``python`` file, they can do so without inadvertently deleting the
  previously installed binary.

> * Even as the Python 2 interpreter becomes less common, scripts should

Reviewer 2: The "should" wording here doesn't match the revised weaker wording above. I'd suggest:

> Even as the Python 2 interpreter becomes less common, it remains reasonable for scripts to continue to use the `python` convention, rather than just `python`. In the event ...

Author: use the python convention, rather than just python? I don't quite understand that bit.

Reviewer 2: Sorry, typo - I left out a critical "3" in the original version of the comment :)

  versa. That way, if a sysadmin does decide to replace the installed
  ``python`` file, they can do so without inadvertently deleting the
  previously installed binary.
* Even as the Python 2 interpreter becomes less common, it remains reasonable
  for scripts to continue to use the ``python3`` convention, rather than just

>   ``python``. In the event of a future Python 4.0 release, distributors may

Reviewer: I'd suggest leaving out any mention of Python 4.  1) it can create FUD; 2) it's entirely vaporware; 3) we don't even know whether there *will* be a Python 4 or what it will look like.  We can always add a Python 4 recommendation when the time comes.

Reviewer 3: Agreed. Good catch @warsaw
